engine/resources/exec.go

```diff
@@ -277,6 +309,16 @@ func (obj *ExecRes) CheckApply(apply bool) (bool, error) {
 		}
 		cmd := exec.Command(cmdName, cmdArgs...)
 		cmd.Dir = obj.IfCwd // run program in pwd if ""
+
+		//TODO: verify with James
+		cmdEnv := make([]string, len(obj.Env))
+		i := 0
+		for key := range obj.Env {
```

Comment:
Look up the map iteration syntax, don't use a counter like `i`. And we'll want to make this deterministic somehow... I'll let you figure it out. Do you know why it isn't deterministic and why it should be?

Reply:
I know that the order of map traversal isn't deterministic in this case. Because of that the end result isn't deterministic because duplicate definitions of a variable aren't assured to be in the same order, possibly leading to an incorrect final definition.
I see the problem, I'll find a solution.

Reply:
Wrong, a map (in golang or mcl) can't have duplicate keys, so _that's_ not the problem. But the list we build will be in different orders which can make debugging or other things tricky. IOW a heisenbug! We never want to cause those. (Eg: "i ran the code and it did X, now it did Y... what's going on? *brain explodes*)

Reply:
So we have to make sure that all of the variables are processed in the order they were written, right? For example we couldn't just sort them alphabetically and process them in that order.

Reply:
I can't think of how we would do that in order just by using a map. I'm assuming that a map is still the best option because you confirmed that to me at one point, but I fail to think of a non-hacky solution

Reply:
> For example we couldn't just sort them alphabetically and process them in that order.

Why not? Of course we can. The solution is to first pull all the keys out of the map into a new list. Then sort that list. Then loop through that list (now deterministic) and read the values out from the map in that order. TADA =D I'll let you write the code for this. Max 6 lines.

Reply:
Okay, I believe I've implemented the necessary changes for this part correctly. I'll submit a new PR in a second and let you take a look.